### ISSUE TITLE
log4cplus waits forever in progschj::~ThreadPool when used in windows DLL

### ISSUE THREAD

**Issue by tbuchloh** [log4cplus/log4cplus#497]
Hi* ,

I want to migrate a windows DLL that uses log4cplus for internal logging, and worked fine for log4cplus 1.2.1 and MSVC2015, to log4cplus 2.0.5 and MSVC2019. 

Unfortunately now my test cases get stuck at process termination.

INITIAL STATE
-------------

1. use log4cplus 2.0.5 source package (https://github.com/log4cplus/log4cplus/archive/REL_2_0_5.zip)
2. MSVC 2019 is installed (MSVC 19.24.28314.0)
3. CMake 3.15.3 installed

TEST STEPS
----------

1. build a library that uses log4cplus for internal logging (linked statically)
2. initialize and deinitialize log4cplus in libraries DLLMain method
3. execute a log statement in library method for demonstration purpose
4. invoke library method from an executable, that is not using log4cplus for itself (so no initialization/deinitialization takes place here).
5. execute executable =>
  5.1. DLLMain initializes log4cplus
  5.2. executable invokes library method and exits normally
  5.2. DLLMain deinitializes log4cplus
  5.3. log4cplus shuts down threadpool
  5.4. ~ThreadPool() waits forever, i. e. process never terminates

EXPECTED OUTCOME
----------------

I would expect that the process will terminate at the end of main() and that destruction will not wait forever on a condition.

OBSERVATIONS
------------

1. it does not matter if log4cplus is linked statically or dynamically.
2. it does not matter if I use log4cplus::initialize()/log4cplus::deinitialize() or a log4cplus::Initializer instance on the heap. I cannot put the initializer into main() because the calling application must not know anything about the DLLs internals.
3. an equivalent behaviour is observed, when DLLMain does not deinitialize explicitly.
4. it does not matter if log4cplus::deinitialize() is called only in case of "lpReserved == NULL" as documentation prescribes (https://docs.microsoft.com/en-us/windows/win32/dlls/dllmain).
5. PropertyConfiguration has no effect.

SAMPLE PROGRAM
--------------

1. see attachment [log4cplus_sample.zip](https://github.com/log4cplus/log4cplus/files/6022459/log4cplus_sample.zip)
2. place log4cplus 2.0.5 source package in subfolder src-pkgs
3. open "x64 Native Tools Command Prompt for VS 2019" and build
```
cd log4cplus_sample
cmake -G"NMake Makefiles" -B _build_static -DBUILD_SHARED_LIBS=FALSE
cmake --build _build_static --target package
```
4. extract package and run "sample_exe". The "sample_exe" will never return.
```
2021-02-22 12:11:47.841Z [     18720] INFO  | dllmain         |   | dllmain.cpp(32) : int __cdecl DllMain(struct HINSTANCE__ *,unsigned long,void *) -- Hello log4cplus!
2021-02-22 12:11:47.852Z [     18720] INFO  | sample_dll      |   | sample_dll.cpp(11) : int __cdecl sample_dll::foo(void) -- foo called!
2021-02-22 12:11:47.852Z [     18720] INFO  | dllmain         |   | dllmain.cpp(38) : int __cdecl DllMain(struct HINSTANCE__ *,unsigned long,void *) -- Goodbye log4cplus!
```
5. run in debugger with your favorite IDE. log4cplus::deinitialize() is called as expected. progschj::ThreadPool::~ThreadPool will wait forever.
```
1  ZwWaitForAlertByThreadId                                                               ntdll                0x7fff106f2a94 
2  RtlSleepConditionVariableSRW                                                           ntdll                0x7fff106a83d9 
3  SleepConditionVariableSRW                                                              KERNELBASE           0x7fff0c88f44d 
4  _crtSetThreadpoolWait                                                                  MSVCP140D            0x7ffeb1738b14 
5  std::_Winerror_message                                                                 MSVCP140D            0x7ffeb16f2406 
6  std::_Winerror_message                                                                 MSVCP140D            0x7ffeb16f2303 
7  std::_Winerror_message                                                                 MSVCP140D            0x7ffeb16f2088 
8  Cnd_wait                                                                               MSVCP140D            0x7ffeb16f25c0 
9  std::condition_variable::wait                                                          mutex           599  0x7ffeaf770ae4 
10 std::condition_variable::wait<<lambda_017a13c3d9b62cc3bfe66e5e914f5053>>               mutex           606  0x7ffeaf6efa78 
11 progschj::ThreadPool::~ThreadPool                                                      ThreadPool.h    157  0x7ffeaf712f3c 
12 progschj::ThreadPool::`scalar deleting destructor'                                     sample_dll           0x7ffeaf7279ec 
13 std::default_delete<progschj::ThreadPool>::operator()                                  memory          1765 0x7ffeaf720b84 
14 std::unique_ptr<progschj::ThreadPool,std::default_delete<progschj::ThreadPool>>::reset memory          1908 0x7ffeaf769cde 
15 log4cplus::shutdownThreadPool                                                          global-init.cxx 337  0x7ffeaf69e2a8 
16 log4cplus::deinitialize                                                                global-init.cxx 556  0x7ffeaf67af6a 
17 DllMain                                                                                dllmain.cpp     39   0x7ffeaf675127 
18 dllmain_dispatch                                                                       dll_dllmain.cpp 281  0x7ffeaf81ab98 
19 _DllMainCRTStartup                                                                     dll_dllmain.cpp 335  0x7ffeaf81ad71 
20 RtlAnsiStringToUnicodeString                                                           ntdll                0x7fff10675c37 
21 LdrShutdownProcess                                                                     ntdll                0x7fff1067911c 
22 RtlExitUserProcess                                                                     ntdll                0x7fff106bdf6d 
23 ExitProcess                                                                            KERNEL32             0x7fff0dc1d62a 
24 wassert                                                                                ucrtbased            0x7ffeb184467a 
25 wassert                                                                                ucrtbased            0x7ffeb1844629 
26 exit                                                                                   ucrtbased            0x7ffeb18449a6 
27 __scrt_common_main_seh                                                                 exe_common.inl  297  0x7ff6992b1787 
28 __scrt_common_main                                                                     exe_common.inl  331  0x7ff6992b162e 
29 mainCRTStartup                                                                         exe_main.cpp    17   0x7ff6992b1959 
30 BaseThreadInitThunk                                                                    KERNEL32             0x7fff0dc17974 
31 RtlUserThreadStart                                                                     ntdll                0x7fff106aa0b1 
```

Thanks a lot for your contribution to this excellent project!

Cheers
Tobias

**Comment by tbuchloh** [comment]
Maybe there is some relationship to #454

**Comment by tbuchloh** [comment]
Well, the patch described in #454 works here as well. Are there any plans to release a 2.0.6 containing this fix?

**Comment by wilx** [comment]
I have published 2.0.6. You can now disable the thread pool entirely. This should help you as workaround this issues with the shutdown.

**Comment by tbuchloh** [comment]
Thank you. That was fast.

Yes, my problems disappeared with 2.0.6. I need to run some more tests, but everything looks very promising now.
